### PR TITLE
Fix migration of VM with volume on Ubuntu

### PR DESCRIPTION
**Pull request by Pearl1594**
### Description

This PR attempts to fix: https://github.com/apache/cloudstack/issues/6114

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by nvazquez** [comment]
@blueorangutan package

**Comment by nvazquez** [comment]
@blueorangutan test ubuntu20 kvm-ubuntu20

**Comment by weizhouapache** [comment]
@Pearl1594 @nvazquez @sureshanaparti 
this sounds like a critical issue in 4.16 as well
(it seems this operation is not supported in centos7 with stock qemu)

**Comment by weizhouapache** [comment]
@Pearl1594 can you re-target to 4.16 ?

**Comment by Pearl1594** [comment]
@blueorangutan test ubuntu20 kvm-ubuntu20

File: engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java [hunk 1]
```diff
@@ -82,10 +82,13 @@
      */
     @Override
     protected StrategyPriority internalCanHandle(Map<VolumeInfo, DataStore> volumeMap, Host srcHost, Host destHost) {
-        if (super.internalCanHandle(volumeMap, srcHost, destHost) != StrategyPriority.CANT_HANDLE
-                || canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) != StrategyPriority.CANT_HANDLE) {
+        if (super.internalCanHandle(volumeMap, srcHost, destHost) != StrategyPriority.CANT_HANDLE) {
             return StrategyPriority.CANT_HANDLE;
         }
+        if (super.internalCanHandle(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE &&
```
Comment (sureshanaparti) [inline engine/storage/datamotion/src/main/java/org/apache/cloudstack/storage/motion/KvmNonManagedStorageDataMotionStrategy.java:88]: @Pearl1594 this check may not be required, it returns true always.

```suggestion
       else if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) != StrategyPriority.CANT_HANDLE) {
```

**Comment by weizhouapache** [comment]
the issue seems to be caused by #5410 
@GabrielBrascher could you please review this PR ?

@Pearl1594 I am wondering if this is better

```
        if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
            return StrategyPriority.CANT_HANDLE;
        }
```

**Comment by Pearl1594** [comment]
I am not sure if it would be correct @weizhouapache (I may be wrong), but based on the 4.15.2 code , it looks like if the strategy returned by `canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost` is CANT_HANDLE, then there are some subsequent checks done before returning can't handle

```
if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
                Set<VolumeInfo> volumeInfoSet = volumeMap.keySet();

                for (VolumeInfo volumeInfo : volumeInfoSet) {
                    StoragePoolVO storagePoolVO = _storagePoolDao.findById(volumeInfo.getPoolId());
                    if (storagePoolVO.getPoolType() != StoragePoolType.Filesystem && storagePoolVO.getPoolType() != StoragePoolType.NetworkFilesystem) {
                        return StrategyPriority.CANT_HANDLE;
                    }
                }
            }
```

**Comment by weizhouapache** [comment]
> I am not sure if it would be correct @weizhouapache (I may be wrong), but based on the 4.15.2 code , it looks like if the strategy returned by `canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost` is CANT_HANDLE, then there are some subsequent checks done before returning can't handle
> 
> ```
> if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
>                 Set<VolumeInfo> volumeInfoSet = volumeMap.keySet();
> 
>                 for (VolumeInfo volumeInfo : volumeInfoSet) {
>                     StoragePoolVO storagePoolVO = _storagePoolDao.findById(volumeInfo.getPoolId());
>                     if (storagePoolVO.getPoolType() != StoragePoolType.Filesystem && storagePoolVO.getPoolType() != StoragePoolType.NetworkFilesystem) {
>                         return StrategyPriority.CANT_HANDLE;
>                     }
>                 }
>             }
> ```

@Pearl1594 
OK, so the logic in this PR looks like same as what before #5410 , right ?

**Comment by GabrielBrascher** [comment]
@weizhouapache thanks for pinging me.
I will double-check the file history to get some context and remember what changed over time and why.
Thanks for stepping up with this PR, @Pearl1594!

I believe that most of the work I've done on that PR was related to extracting a few validations added by @nvazquez in https://github.com/apache/cloudstack/commit/0fbf5006b87b6c5bc6188e3e1dd226a98c3bb453 and align with the workflow in case of `Local volume` +` attached data disks`. Might be good if @nvazquez can join us and double-check it as well.

I have some meetings and tasks to run. Get back as soon as I can to refresh my memory and re-debug this flow.

Regarding @Pearl1594`s comment in https://github.com/apache/cloudstack/pull/6116#issuecomment-1068867617
@Pearl1594 I believe that you are right indeed. There are some special cases to be handled, with regards to being NFS migration or a local storage migration, which led to the subsequent checks you've referenced.

**Comment by Pearl1594** [comment]
> > I am not sure if it would be correct @weizhouapache (I may be wrong), but based on the 4.15.2 code , it looks like if the strategy returned by `canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost` is CANT_HANDLE, then there are some subsequent checks done before returning can't handle
> > ```
> > if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
> >                 Set<VolumeInfo> volumeInfoSet = volumeMap.keySet();
> > 
> >                 for (VolumeInfo volumeInfo : volumeInfoSet) {
> >                     StoragePoolVO storagePoolVO = _storagePoolDao.findById(volumeInfo.getPoolId());
> >                     if (storagePoolVO.getPoolType() != StoragePoolType.Filesystem && storagePoolVO.getPoolType() != StoragePoolType.NetworkFilesystem) {
> >                         return StrategyPriority.CANT_HANDLE;
> >                     }
> >                 }
> >             }
> > ```
> 
> @Pearl1594 OK, so the logic in this PR looks like same as what before #5410 , right ?

Yes @weizhouapache

**Comment by weizhouapache** [comment]
> > > I am not sure if it would be correct @weizhouapache (I may be wrong), but based on the 4.15.2 code , it looks like if the strategy returned by `canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost` is CANT_HANDLE, then there are some subsequent checks done before returning can't handle
> > > ```
> > > if (canHandleKVMNonManagedLiveNFSStorageMigration(volumeMap, srcHost, destHost) == StrategyPriority.CANT_HANDLE) {
> > >                 Set<VolumeInfo> volumeInfoSet = volumeMap.keySet();
> > > 
> > >                 for (VolumeInfo volumeInfo : volumeInfoSet) {
> > >                     StoragePoolVO storagePoolVO = _storagePoolDao.findById(volumeInfo.getPoolId());
> > >                     if (storagePoolVO.getPoolType() != StoragePoolType.Filesystem && storagePoolVO.getPoolType() != StoragePoolType.NetworkFilesystem) {
> > >                         return StrategyPriority.CANT_HANDLE;
> > >                     }
> > >                 }
> > >             }
> > > ```
> > 
> > 
> > @Pearl1594 OK, so the logic in this PR looks like same as what before #5410 , right ?
> 
> Yes @weizhouapache

@Pearl1594 
OK, looks good to me. let's wait for the smoke test results

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by Pearl1594** [comment]
@blueorangutan test ubuntu20 kvm-ubuntu20

**Comment by Pearl1594** [comment]
@blueorangutan package

**Comment by Pearl1594** [comment]
@blueorangutan test ubuntu20 kvm-ubuntu20

**Review by GabrielBrascher — APPROVED** [review]
I will run some manual tests ASAP.
Waiting also for the smoke test results.
But code LGTM

**Comment by GabrielBrascher** [comment]
@Pearl1594 @weizhouapache tested and it is working fine when migrating VMs with `KVM` + `Root on Local`.
In case tests are all good, I am LGTM in having this one merged.

**Review by nvazquez — APPROVED** [review]
LGTM thanks @Pearl1594 and @GabrielBrascher

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
Tests on ubuntu are fixed, started a new round of tests for centos to ensure no regressions are added - can be merged after that

**Review by rohityadavcloud — APPROVED** [review]

**Comment by Pearl1594** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
@blueorangutan test

**Comment by nvazquez** [comment]
Merging based on approvals and test results on ubuntu and KVM (error not related to the PR)

**Pull request merged**